File: killrvideo-service-users/src/main/java/com/killrvideo/service/user/repository/UserRepository.java
package com.killrvideo.service.user.repository;

import com.datastax.oss.driver.api.core.CqlSession;
import com.killrvideo.service.user.dao.UserCredentialsDao;
import com.killrvideo.service.user.dao.UserDao;
import com.killrvideo.service.user.dao.UserMapper;
import com.killrvideo.service.user.dto.User;
import com.killrvideo.service.user.dto.UserCredentials;
import com.killrvideo.dse.utils.MappedAsyncPagingIterableUtils;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

/**
 * Handling user.
 *
 * @author DataStax Developer Advocates Team
 */
@Repository
public class UserRepository {
    private UserDao userDao;
    private UserCredentialsDao userCredentialsDao;

    public UserRepository(CqlSession session) {
        UserMapper mapper = UserMapper.builder(session).build();
        this.userDao = mapper.getUserDao();
        this.userCredentialsDao = mapper.getUserCredentialsDao();
    }

    /**
     * Create user Asynchronously composing things. (with Mappers)
     *
     * @param user           user Management
     * @param hashedPassword hashed Password
     * @return
     */
    public CompletableFuture<User> createUserAsync(User user, String hashedPassword) {
        return userCredentialsDao.insert(UserCredentials.from(user, hashedPassword))
                .thenCompose(rs -> userDao.insert(user));
    }

    /**
     * Get user Credentials
     *
     * @param email
     * @return
     */
    public CompletableFuture<UserCredentials> getUserCredentialAsync(String email) {
        return userCredentialsDao.getUserCredential(email);
    }

    /**
     * Retrieve user profiles.
     *
     * @param userids
     * @return
     */
    public CompletableFuture<List<User>> getUserProfilesAsync(List<UUID> userids) {
        return userDao.getUserProfiles(userids)
                .thenApply(MappedAsyncPagingIterableUtils::all);
    }
}

File: killrvideo-service-users/src/main/kotlin/com/killrvideo/service/user/repository/UserRepository.kt
package com.killrvideo.service.user.repository

import com.killrvideo.dse.utils.MappedAsyncPagingIterableExtensions.all
import com.killrvideo.service.user.dao.UserCredentialsDao
import com.killrvideo.service.user.dao.UserDao
import com.killrvideo.service.user.dao.UserMapper
import com.killrvideo.service.user.dto.User
import com.killrvideo.service.user.dto.UserCredentials
import io.grpc.Status
import kotlinx.coroutines.future.await
import mu.KotlinLogging
import org.springframework.stereotype.Repository
import java.util.*

/**
 * Handling user.
 *
 * @author DataStax Developer Advocates Team
 */
@Repository
open class UserRepository(mapper: UserMapper) {
    private val logger = KotlinLogging.logger {  }
    private val userDao: UserDao = mapper.userDao
    private val userCredentialsDao: UserCredentialsDao = mapper.userCredentialsDao

    /**
     * Create user Asynchronously composing things. (with Mappers)
     *
     * @param user           user Management
     * @param hashedPassword hashed Password
     * @return
     */
    suspend fun createUserAsync(user: User, hashedPassword: String): Boolean {
        val userCredentials: UserCredentials = UserCredentials.from(user, hashedPassword)
        return userCredentialsDao.insert(userCredentials)
            .thenApply { success ->
                duplicateCheck(userCredentials, !success)
                success
            }
            .thenCompose { _ -> userDao.insert(user) }
            .await()
    }

    private fun duplicateCheck(userCredentials: UserCredentials, duplicate: Boolean) {
        if (duplicate) {
            val errMsg = "Exception creating user because it already exists with email ${userCredentials.email}"
            logger.error(errMsg)
            throw Status.ALREADY_EXISTS.withDescription(errMsg).asRuntimeException()
        }
    }

    /**
     * Get user Credentials
     *
     * @param email
     * @return
     */
    suspend fun getUserCredentialAsync(email: String): UserCredentials? =
        userCredentialsDao.getUserCredential(email).await()

    /**
     * Retrieve user profiles.
     *
     * @param userids
     * @return
     */
    suspend fun getUserProfilesAsync(userids: List<UUID>): List<User> =
        userDao.getUserProfiles(userids)
            .thenApply { it.all() }.await()
}
